[RND-425] OAuth2 Refactors and Documentation (#146)
Ed-Fi-Exchange-OSS/Meadowlark: docs/OAUTH2.md

to integrate with third party providers.

One implication: the API application should not need to know the signing key;
only the OAuth2 provider will have that. Thus the client application must use
the token introspection endpoint to validate a received token. Validated tokens
are cached so that subsequent requests do not need to revalidate them, except
for confirming the expiration date. Tokens remain in the cache by default for up
to 5 minutes; if there are 1000 (configurable) tokens then the oldest ones are
vacated more quickly to make room for more tokens.

### Client Configuration

These environment variables are on the Meadowlark client side; when we get to
the point of having third party OAuth providers, these settings will still be
relevant, whereas the settings above will not be needed.

| Variable                                      | Default Value     | Purpose                                               |
| --------------------------------------------- | ----------------- | ----------------------------------------------------- |
| OAUTH_CLIENT_PROVIDED_TOKEN_CACHE_TTL         | 300000 ms (5 min) | Length of time to accept a token without revalidating |

> | OAUTH_CLIENT_PROVIDED_TOKEN_CACHE_MAX_ENTRIES | 1000              | Max number of validated tokens to cache               |

Reviewer: What about the four env vars here:

https://github.com/Ed-Fi-Exchange-OSS/Meadowlark/blob/707bc47367a4f742a544a4b42372d458b68bd39f/Meadowlark-js/packages/meadowlark-core/src/middleware/OAuthFetch.ts#L12

  "expires_in": 1659374228,
  "refresh_token": "not available"
}
```

### Token Introspection

This verb and endpoint supports the concept of [token
introspection](https://datatracker.ietf.org/doc/html/rfc7662). The Meadowlark
API as a client calls this endpoint to validate the received token.

Sample request:

```none
POST /oauth/verify
Authorization: bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI....

token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI....
```

Note that only form-urlencoded requests are accepted, per the specification.

Only an authenticated client can post to the `/oauth/verify` endpoint. A client
with the "admin" role can post _any_ token, whereas other clients can only post
their _own_ token for introspection.

#### Valid Token Response

Contains the "plain text" of the JWT, with the additional `active` attribute
showing that the token is not expired and has not otherwise been invalidated.

```json
{
  "active": true,
  "client_id": "fbf739c4-fb86-4f03-a477-91af51cc46f2",
  "sub": "client name",
  "aud": "ed-fi-meadowlark",
  "iss": "ed-fi-meadowlark",
  "exp": 1659374285,
  "iat": 1659374285,
  "roles": [
    "vendor"
  ]
}
```

#### Invalid Token Response

> If the token is expired, or has been invalidated (for example, the client's
> access has been revoked), then the response will look like:

Author: ```suggestion
If the token is expired, or has been invalidated (for example, the client
has been deactivated), then the response will look like:
```

```json
{
  "active": false
}
```

### Client Credential Management

Only clients with the "admin" role may access these endpoints.

- `GET /oauth/client`
- `GET /oauth/client/{client id}`
- `POST /oauth/client`

  ```json
  {
    "clientName": "Hometown SIS",
    "roles": [
      "vendor"
    ]
  }
  ```

- `PUT /oauth/client/{client id}`

  ```json

>   {
>     "client_id": "{client id}",
>     "clientName": "Hometown SIS",

Author: ```suggestion
  {
    "active": true,
    "client_id": "{client id}",
    "clientName": "Hometown SIS",
```

    "roles": [
      "vendor"
    ]
  }
  ```

> - `DELETE /oauth/client/{client id}` - deactivates, does not remove the client

Author: ```suggestion
```
